internal/bzlmod/go_deps.bzl

```diff
@@ -271,6 +277,77 @@ _go_repository_config = repository_rule(
+    """
+
+    if not previous:
+        # no previous module, so no possible error
+        return
+
+    if not previous or version == previous.version:
+        # version is the same, skip because we won't error
+        return
+
+    # When using go.work, duplicate dependency versions are possible.
+    # This can cause issues, so we fail with a hopefully actionable error.
+    current_label = module_tag._parent_label
+
+    previous_label = previous.module_tag._parent_label
+
+    corrective_measure = None
+    default_corrective_mesasure = "To correct this:\n 1. manually update: all go.mod files to ensure the versions of '{}' are the same.\n 2. in the folder where you made changes to run: go mod tidy\n 3. run: go work sync.".format(module_tag.path)
+
+    if previous.version[0] == version[0] or str(current_label).endswith("go.work") or str(previous_label).endswith("go.work"):
+        corrective_measure = default_corrective_mesasure
+    else:
+        label = module_name_to_go_dot_mod_label.get(module_tag.path)
+
+        if label:
+            # if the label is present that means the module_tag is of a go.mod file, which means the correct action may be different.
+
+            # if the duplicate module in question is provided by go.work use statement then only manual intervention can fix it
+            # from_file_tags on go_work represents use statements in the go.work file
+            for from_file_tags in [go_work.from_file_tags for go_work in go_works]:
+                for from_file_tag in from_file_tags:
+                    if from_file_tag.go_mod == label:
+                        corrective_measure = default_corrective_mesasure
+                        break
+        elif previous.module_tag.indirect or module_tag.indirect:
+            # if the dependency indirect, the user will need to manually update go.mod, run go mod tidy in that directory and then run go work sync
+            corrective_measure = default_corrective_mesasure
+        else:
+            # TODO: if the version are v0.8.0 and v0.17.0 go work sync wont work
+            # ensure the corrective measure describes this. Maybe this is limited to indirect dependencies
+            corrective_measure = "To correct this, run:\n 1. go work sync."
+
+    message = "Multiple versions of {} found:\n - {} contains: {}\n - {} contains {}.\n{}".format(module_tag.path, current_label, semver.humanize_comparable(version), previous_label, semver.humanize_comparable(previous.version), corrective_measure)
+
+    if fail_or_warn:
+        fail(message)
+    else:
+        print(message)
+
+def _fail_if_not_root(module, from_file_tag):
+    if module.is_root != True:
+        fail("go_deps.from_file(go_work = '{}') tag can only be used from a root module but: '{}' is not a root module.".format(from_file_tag.go_work, module.name))
+
+def _fail_if_invalid_from_file_usage(from_file_tag):
+    if (
+        (from_file_tag.go_work == None and from_file_tag.go_mod == None) and
```

Comment:
`bool(from_file_tag.go_work) == bool(from_file_tag.go_mod)` could be considered simpler (your call)

Reply:
That is much nicer!